pkg/storage/bolt_test.go

```diff
@@ -1,14 +1,16 @@
 package storage
 
 import (
-	"github.com/goccy/go-json"
-	"github.com/stretchr/testify/assert"
 	"os"
 	"testing"
+
+	"github.com/goccy/go-json"
+	"github.com/stretchr/testify/assert"
 )
 
 func TestBoltDB(t *testing.T) {
-	db, err := NewBoltDBWithFile("test.db")
+	db := &BoltDB{}
+	err := db.Init("test.db")
```

Comment:
I'm not sure I see the benefit in this type of pattern

i don't think users should be instantiating `&BoltDB{}` directly - a method would be preferable as there was before

Reply:
let's see the code blew:

```go
// NewStorage creates a new storage provider based on the input
func NewStorage(storageProvider Storage) (ServiceStorage, error) {
	switch storageProvider {
	case Bolt:
		return NewBoltDB()
	default:
		return nil, fmt.Errorf("unsupported storage provider: %s", storageProvider)
	}
}
```
think about if we  don't use current pattern.  every time we add an implementation, we should add one case to `NewStorage`. and this way also not support 3rd party register.

but, use current pattern, Whatever new implementation added or  3rd party implementation registed, `NewStorage` code will not change.  just like blew:

```go
// NewStorage creates a new storage provider and provider specify param based on the input
func NewStorage(storageProvider Storage, param interface{}) (ServiceStorage, error) {
	impl := GetStorage(storageProvider)
	if impl == nil {
		return impl, fmt.Errorf("unsupported storage provider: %s", storageProvider)
	}
	err := impl.Init(param)
	return impl, err
}
```

>> i don't think users should be instantiating &BoltDB{} directly - a method would be preferable as there was before

if we use functions like NewBoltDB , different implementations will use different NewXXXX function so we can not create and set up implementations in NewStorage automaticlly (i also add an uniform standard Init(interface{}) into ServiceStorage, so we can do set up with params for implementation ).

```go
...
err := impl.Init(param)
...
```

Reply:
Thanks @suutaku I think this is the key of my gap in understanding.

We can make the pattern more generic as you have here, but practically - how will a 3rd party option become registered?

I don't foresee anyone using this project as a library. To modify it they would need to fork and make changes. With that burden, does this change really make much simpler?

From another perspective I can see not wanting to have any DB be the "default" for the service. I think that is a valuable change (though may already be possible?).

Reply:
I have the same question as Gabe. When adding a new storage, is it simpler to add an import with side effects, or to add a new case in the switch statement? 

Additionally, related to `init` functions. How would we enforce that developers do registration of the storage within the module (or package) they are adding?

Reply:
@decentralgabe @andresuribe87 Thanks to replay me! About questions:
> The case of use `ssi-service` as a library

 `ssi-service` can be a part of another service,  currently, we can use `micro-service way` to do that, but i think we also should support `import way`.
A simple example, i have a repo, named `my-repo`, i have a new implementation of storage (can be a 3rd party repo, or just a single .go file in `my-repo`).  at some where of my code, I import `ssi-service`, called `RegisterStorage`, use `ssi-service` as a instance and everything is done.

> how will a 3rd party option become registered?

you mean options in configure ?  When developers call `RegisterStorage` register 3rd party storage, option is not registered.  developers should modify configure file (the `StorageParam` or `StorageOptions` tag) and the `NewStorage` called, storage specify 3rd party options will applied according to `Init(interface{})` function's implementation. in `Init(interface{})` function developer can do everything they want to do with the option(s).

>  How would we enforce that developers do registration of the storage within the module (or package) they are adding?

We cannot force developer do anything. we can just tell developers, there was a simple way to use our code: import it, call `RegiserStorage` and Ok.  in fact,  they can call `RegiserStorage` any place the want, the `init()` function just an option.

Note that, in this approach, developers will not modify any single code in `ssi-service`.

Reply:
@decentralgabe I agree with you. the "default" (currently it is `bolt`) just make `ssi-service` working.  
> From another perspective I can see not wanting to have any DB be the "default" for the service. I think that is a valuable change (though may already be possible?).

Reply:
> at some where of my code, I import ssi-service, called RegisterStorage, use ssi-service as a instance and everything is done.

Gotcha! I now understand better what you're trying to accomplish, and the registry pattern certainly does it well. 

That said, I would prefer explicit dependency injection so developers have an easier time tracking down what instances are being used. In this case, I propose the following alternative. 

1. Modify the `ServicesConfig` struct to have a method called `ServiceStorage`. 
2. The `ServiceStorage` would have a default implementation that returns the bold storage.

Devs can then inject whatever `ServiceStorage` implementation they like, without the need of modifying the underlying code. 

The benefits of this approach is that we don't use the `init` mechanism, which I've seen in the past is difficult to understand.

Reply:
yes, that's also a good solution! in face, we can also don't use the init mechanism, just call `RegisterStorage` some where. 
and, I would prefer  keep the `availableStorage map[Storage]StorageService`, because with it, we can event switch storage in dynamic(the case of use multiple storage).
so the flow can be: setting multiple storage string in configure file ->  inject multiple storage to service -> make a switch logic -> use one of a storage.

Reply:
agree with this approach!

Reply:
Let's make `StorageService` to manage `Storage` implementations first? Use `ServicesConfig` do that will be a little confused? in my opinion, the configure only manage the configure files, it' s text work. 
also some day, we need configure watching the conf files if it was modified,  that's one of the way we can change (switch) settings dynamically.